### PR TITLE
Add implementation of PML

### PR DESCRIPTION
**Pull request by sbastrakov**
I've now put my changes on top of the current dev. There are still some leftovers from the previous efforts, and, importantly, the kernel is still not changed to work in the internal area vs. the guard as originally planned. So still WIP, but hopefully just for 1-2 more days.

I'll now close the older PR #2863 .

To enable the new solver, use `YeePML` as the field solver in `fieldSolver.param`. All the PML-related parameters are in `pml.param` (default values should normally be reasonable), parameters of the old exponential damping absorber should still be specified so that the code compiles, but are not actually used. Otherwise, PML behaves exactly the same as the old absorber, in terms of absorbing area location (inside the simulation area, near the outer border) and interaction with lasers and moving window.
Known limitations:

- When PML is enabled, for each grid node on the whole domain where are 12 additional `float_X` values, so memory requirements for field data on both host and device are increased by about 2 times (particles' data is not affected, ofc). This issue will be addressed in a following PR.
- That additional values are not written to checkpoints. It is technically possible to load from a checkpoint, but field dynamics in the PML area will be changed and reflections introduced, so not worthy for meaningful simulations. This will be addressed shortly when absorbers become plugins in a following PR.
- Now field solver kernels are much more computationally intensive and use way more registers. However, in general PIC scale this should be negligible, as most time is spent on the other stages, and the field solver kernels are also interleaved with asynchronous MPI communications. Even in my tests with no particles, there is no time difference between the `Yee` + old absorber and new `YeePML`.

### CODE REVIEW

File: include/picongpu/fields/MaxwellSolver/Yee/Curl.hpp [hunk 1]
```diff
@@ -35,9 +35,10 @@ namespace maxwellSolver
 namespace yee
 {
 
-    template< typename  Difference >
+    template< typename T_Difference >
```
Comment (psychocoderHPC) [inline include/picongpu/fields/MaxwellSolver/Yee/Curl.hpp]: Even if this change is small please provide this change as a separate PR because it is independent to this PR.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+            }
+
+            //! Get a reference to (full) field B
+            picongpu::FieldB& getFieldB( )
+            {
+                return *( fieldB.get() );
+            }
+
+            /**
+             * Propagate B values in the given area by half a time step.
```
Comment (psychocoderHPC) [inline include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp]: Please put the brief description to the first line `/**`

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+                 * at a border we set the corresponding thickness to 0.
+                 */
+                auto & movingWindow = MovingWindow::getInstance( );
+                auto const numSlides = movingWindow.getSlideCounter( currentStep );
+                auto const numExchanges = NumberOfExchanges< simDim >::value;
+                auto const communicationMask = Environment< simDim >::get( ).GridController( ).getCommunicationMask( );
+                Thickness localThickness = globalSize;
+                for( auto exchange = 1; exchange < numExchanges; ++exchange )
+                {
+                    // Ignore directions except left, right, top, bottom, back, front
```
Comment (psychocoderHPC) [inline include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp]: I would rephrase it to `Skip directions ...`

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+                    {
+                        return;
+                    }
+                }
+
+                constexpr float_X c2 = SPEED_OF_LIGHT * SPEED_OF_LIGHT;
+                constexpr float_X dt = DELTA_T;
+
+                // Shift indexes so that 0 is start of the local domain without guard
+                const float3_X sigmaIdx(
```
Comment (psychocoderHPC) [inline include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel]: You can also write 

```C++
const float3_X sigmaIdx(
    precisionCast<float_X>(cellIdx - numGuardCells)
);
```

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+                    cellIdx.x() - numGuardCells.x(),
+                    cellIdx.y() - numGuardCells.y(),
+                    cellIdx.z() - numGuardCells.z()
+                );
+                const float3_X sigma = detail::getSigma( sigmaIdx, parameters, numLocalDomainCells );
+                const float3_X damping(
```
Comment (psychocoderHPC) [inline include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel]: we should define the dotwise `exp` function for pmacc vectors and then write 

```C++
const float3_X damping( 
     math::exp( -sigma * dt )
);
```

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+                    cellIdx.y() - numGuardCells.y(),
+                    cellIdx.z() - numGuardCells.z()
+                );
+                const float3_X sigma = detail::getSigma( sigmaIdx, parameters, numLocalDomainCells );
+                const float3_X damping( 
+                    math::exp( -sigma[0] * dt ),
+                    math::exp( -sigma[1] * dt ),
+                    math::exp( -sigma[2] * dt )
+                );
+                float3_X diff( dt, dt, dt );
```
Comment (psychocoderHPC) [inline include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel]: can be written as `auto diff = float3_X::create( dt );`

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+                for ( uint32_t dim = 0; dim < 3; dim++ )
+                    if ( sigma[ dim ] )
+                        diff[ dim ] = ( 1.0_X - damping[ dim ] ) / sigma[ dim ];
+                ///std::cout << "E: cellIdx = " << cellIdx << ", damping = " << damping << ", diff = " << diff << "\n";
```
Comment (psychocoderHPC) [inline include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel]: note: dead debug code

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+            float_X index,
+            float_X numPMLCellsNegative,
+            float_X numPMLCellsPositive,
+            uint32_t numLocalDomainCells
+        )
+        {
+            float_X result = 0.0;
+            if( index < numPMLCellsNegative )
+                result = ( numPMLCellsNegative - index ) / numPMLCellsNegative;
+            if( index > numLocalDomainCells - numPMLCellsPositive )
```
Comment (psychocoderHPC) [inline include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel]: should be an `else if` since only one path can be valid.

File: include/picongpu/fields/MaxwellSolver/YeePML/Field.tpp [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+namespace picongpu
+{
+namespace fields
+{
+namespace maxwellSolver
+{
+namespace yeePML
+{
+
+Field::Field( MappingDesc cellDescription ) :
```
[the remark this answers is not in the document]
Comment (sbastrakov) [reply include/picongpu/fields/MaxwellSolver/YeePML/Field.tpp]: Thanks, fixed.

**Comment by sbastrakov** [comment]
I made experiments on setups from #2948, including the commented out ones, to compare the absorbing performance of the exponential damping absorber and the new PML implementation. Everything seems to behave as it should. PML provides a much better absorption even with less thickness (as expected). So it should be a strictly better alternative, except additional memory for fields. 

With parameters `ALPHA_MAX_SI` and `KAPPA_MAX` (in `pml.param`) a user can switch between different versions of PML. Alpha > 0 corresponds to convolutional (CPML), zero - to uniaxial (UPML). Having Kappa > 1 means coordinate stretching in the PML area. As the literature and my experiments suggest, there is no option that is clearly superior to others. The default parameters with 8 cells of CPML and no stretching should be generally reasonable.

Attaching the results in Excel and plots separately as pictures.
[PlotsPML.xlsx](https://github.com/ComputationalRadiationPhysics/picongpu/files/3178523/PlotsPML.xlsx)
CPML_vs_UPML_8cells_2D
CPML_vs_UPML_8cells_3D
CPML_vs_UPML_16cells_2D
CPML_vs_UPML_16cells_3D
PML_vs_ExponentialDamping_2D
PML_vs_ExponentialDamping_3D

**Comment by sbastrakov** [comment]
The code is ready for review.

**Comment by PrometheusPi** [comment]
@psychocoderHPC since #2948 depends on this pull request, how is your plan to review this pull request?

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp [hunk 10]
```diff
@@ -0,0 +1,431 @@
...
+     * Absorption is done using convolutional perfectly matched layer (CPML),
+     * implemented according to [Taflove, Hagness].
+     *
+     * This class template is a public interface to be used, e.g. in .param
+     * files and is compatible with other field solvers. Parameters of PML
+     * are taken from pml.param, pml.unitless.
+     *
+     * Enabling this solver results in more memory being used on a device:
+     * 12 additional scalar field values per each grid cell of a local domain.
+     * Another limitation is not full persistency with checkpointing: the
```
Comment (PrometheusPi) [inline include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp]: This is a quite hard limitation. @psychocoderHPC, @ax3l and @steindev what do you think about that? Fore me this limits usability for long simulations where restarts are required. However, usually those are the ones that need better absorbers.

**Comment by psychocoderHPC** [comment]
@PrometheusPi I will review this and merge it and then we can merge #2948

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel [hunk 11]
```diff
@@ -0,0 +1,587 @@
...
+        }
+
+        /** Get absorption parameters: sigma, kappa and alpha at a given cell
+         *
+         * Apply polynomial grading, as described in pml.param.
+         *
+         * @param cellIdx cell index including the guard, can be fractional,
+         * e.g. for halves of cells
+         * @param parameters parameters of PML in the local domain
+         * @param normalizedSigma value of normalized sigma at the cell, output
```
Comment (PrometheusPi) [inline include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel]: Why do you add `, output`? 
Same applies to two lines below.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel [hunk 12]
```diff
@@ -0,0 +1,587 @@
...
+                        psiE.xy = coeffs.b.y( ) * psiE.xy + coeffs.c.y( ) * dBzDy;
+                        psiE.zy = coeffs.b.y( ) * psiE.zy + coeffs.c.y( ) * dBxDy;
+                        psiE.xz = coeffs.b.z( ) * psiE.xz + coeffs.c.z( ) * dByDz;
+                        psiE.yz = coeffs.b.z( ) * psiE.yz + coeffs.c.z( ) * dBxDz;
+
+                        /* [Taflove, Hagness], eq. (7.106) and similar for other
+                         * components. Coefficients Ca, Cb as given in (7.107a,b)
+                         * are general to account for materials, in addition to
+                         * artificial PML absorbing medium. We do not have any
+                         * real material, so in (7.107a,b) have to use
```
Comment (PrometheusPi) [inline include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel]: missing `we`: ... **we** have to ...

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel [hunk 13]
```diff
@@ -0,0 +1,587 @@
...
+                         * sigma(i + 1/2, j, k) = 0 (it is another sigma,
+                         * unrelated to PML), eps(i + 1/2, j, k) = EPS0. Also,
+                         * same as the Yee scheme in PIC, adjust to use B,
```
Comment (PrometheusPi) [inline include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel]: adjust**ed**?

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel [hunk 14]
```diff
@@ -0,0 +1,587 @@
...
+                         * not H, in the right-hand side.
```
Comment (PrometheusPi) [inline include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel]: ~in~ on [see here](https://www.dict.cc/englisch-deutsch/on+the+right+hand+side.html)

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel [hunk 15]
```diff
@@ -0,0 +1,587 @@
...
+        typename T_BlockDescription
+    >
+    struct KernelUpdateBHalf
+    {
+        /** Update the magnetic field by half a time step
+         *
+         * @tparam T_Acc alpaka accelerator type
+         * @tparam T_Mapping mapper functor type
+         * @tparam T_Curl curl functor type
+         * @tparam T_EBox pmacc::DataBox, electric field box type
```
Comment (PrometheusPi) [inline include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel]: I think the extra comma `,`is not needed here - same is true for the line below.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel [hunk 16]
```diff
@@ -0,0 +1,587 @@
...
+                    uint32_t const linearIdx,
+                    uint32_t const
+                )
+                {
+                    constexpr auto halfDt = 0.5_X * DELTA_T;
+                    auto const idxInSuperCell =
+                        DataSpaceOperations< simDim >::template map< SuperCellSize >( linearIdx );
+                    // grid index to process with the current thread
+                    auto const idx = blockBeginIdx + idxInSuperCell;
+                    // with the current Yee grid, a half cell shift needed here
```
Comment (PrometheusPi) [inline include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel]: missing `is`: **is** needed here

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel [hunk 17]
```diff
@@ -0,0 +1,587 @@
...
+                        psiB.yz = coeffs.b.z( ) * psiB.yz + coeffs.c.z( ) * dExDz;
+
+                        /* [Taflove, Hagness], eq. (7.108) and similar for other
+                         * components. Coefficients Da, Db as given in (7.109a,b)
+                         * are general to account for materials, in addition to
+                         * artificial PML absorbing medium. We do not have any
+                         * real material, so in (7.107a,b) have to use
+                         * sigma*(i + 1/2, j, k) = 0 (it is another sigma*,
+                         * unrelated to PML), mue(i + 1/2, j, k) = MUE0. Also,
+                         * same as the Yee scheme in PIC, adjust to use B,
```
Comment (PrometheusPi) [inline include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel]: adjust**ed**? (see comment before)

File: include/picongpu/param/pml.param [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+namespace picongpu
+{
+namespace fields
+{
+namespace maxwellSolver
+{
+namespace yeePML
+{
+
+    /* The parameters in this file are only used if field solver is YeePML.
```
Comment (PrometheusPi) [inline include/picongpu/param/pml.param]: Okay - perhaps: `if the field solver is set to YeePML` or  `if the field solver selected is YeePML`.

File: include/picongpu/param/pml.param [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+     * The original paper on this approach is J.A. Roden, S.D. Gedney.
+     * Convolution PML (CPML): An efficient FDTD implementation of the CFS - PML
+     * for arbitrary media. Microwave and optical technology letters. 27 (5),
+     * 334-339 (2000).
+     * https://doi.org/10.1002/1098-2760(20001205)27:5%3C334::AID-MOP14%3E3.0.CO;2-A
+     * Our implementation based on a more detailed description in section 7.9 of
```
Comment (PrometheusPi) [inline include/picongpu/param/pml.param]: **is** based (see comment before)

File: include/picongpu/param/pml.param [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+
+    /** Thickness of the absorbing layer, in number of cells
+     *
+     * PML is located inside the global simulation area, near the outer borders.
+     * Setting size to 0 results in disabling absorption at the corresponding
+     * boundary. Normally thickness is between 6 and 16 cells, with larger
+     * values providing less reflections.
+     * 8 cells should be good enough for most simulations. There are no
+     * requirements on thickness being a multiple of the supercell size.
+     * It is only required that PML is small enough to fit near-boundary local
```
Comment (PrometheusPi) [inline include/picongpu/param/pml.param]: see comment/question before: This is a bit unclear to me.

File: include/picongpu/fields/MaxwellSolver/YeePML/Parameters.hpp [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+
+        /** Max value of coordinate stretching coefficient
+         *
+         * Unitless.
+         */
+        floatD_X kappaMax;
+
+        /** Max value of complex frequency shift
+         *
+         * Components correspond to directions. Normalized by eps0.
```
[the remark this answers is not in the document]
Comment (sbastrakov) [reply include/picongpu/fields/MaxwellSolver/YeePML/Parameters.hpp]: The complex stretching transformation is given by `x -> kappa_x + sigma_x / (alpha_x + i * omega * eps0)`, so `alpha` actually has this unit. Whether it is a good name is arguable, but this is really standard in the PML literature.

File: include/picongpu/fields/MaxwellSolver/YeePML/Field.hpp [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+
+        virtual void reset(uint32_t currentStep);
+
+        HDINLINE static UnitValueType getUnit();
+
+        /** powers of the 7 base measures
+         *
+         * characterizing the record's unit in SI
+         * (length L, mass M, time T, electric current I,
+         *  thermodynamic temperature theta, amount of substance N,
```
[the remark this answers is not in the document]
Comment (sbastrakov) [reply include/picongpu/fields/MaxwellSolver/YeePML/Field.hpp]: Thanks, will fix.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp [hunk 23]
```diff
@@ -0,0 +1,431 @@
...
+    {
+
+        /** Implementation of Yee + PML solver updates of E and B
+         *
+         * The original paper on this approach is J.A. Roden, S.D. Gedney.
+         * Convolution PML (CPML): An efficient FDTD implementation of the
+         * CFS - PML for arbitrary media. Microwave and optical technology
+         * letters. 27 (5), 334-339 (2000).
+         * https://doi.org/10.1002/1098-2760(20001205)27:5%3C334::AID-MOP14%3E3.0.CO;2-A
+         * Our implementation based on a more detailed description in section
```
[the remark this answers is not in the document]
Comment (sbastrakov) [reply include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp]: Well spotted.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp [hunk 24]
```diff
@@ -0,0 +1,431 @@
...
+                constexpr auto numWorkers = getNumWorkers( );
+                using Kernel = yeePML::KernelUpdateBHalf<
+                    numWorkers,
+                    BlockDescription< CurlE >
+                >;
+                AreaMapper< T_Area > mapper{ cellDescription };
+                /* Note: here it is possible to first check if PML is enabled
+                 * in the local domain at all, and otherwise optimize by calling
+                 * the normal Yee update kernel. We do not do that, as this
+                 * would be fragile wrt future separation of PML into a plugin.
```
[the remark this answers is not in the document]
Comment (sbastrakov) [reply include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp]: I meant wrt == with respect to, will write as the full expression to avoid confusion.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp [hunk 25]
```diff
@@ -0,0 +1,431 @@
...
+            /* PML convolutional field data, defined as in [Taflove, Hagness],
+             * eq. (7.105a,b), and similar for other components
+             */
+            std::shared_ptr< yeePML::FieldE > psiE;
+            std::shared_ptr< yeePML::FieldB > psiB;
+
+            /** Thickness in terms of the global domain.
+             *
+             * We store only global thickness, as the local one can change
+             * during the simulation and so has to be recomputed each time step.
```
[the remark this answers is not in the document]
Comment (sbastrakov) [reply include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp]: WIll fix.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel (hunk 16 above)
Comment (sbastrakov) [reply include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel]: Thanks, will fix.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel (hunk 17 above)
Comment (sbastrakov) [reply include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel]: Agree.

File: include/picongpu/param/pml.param (hunk 18 above)
Comment (sbastrakov) [reply include/picongpu/param/pml.param]: Good point.

File: include/picongpu/param/pml.param (hunk 20 above)
Comment (sbastrakov) [reply include/picongpu/param/pml.param]: Not sure which comment/question do you mean?

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel (hunk 11 above)
Comment (sbastrakov) [reply include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel]: Will change these to the doxygen `@param[out]`.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp [hunk 26]
```diff
@@ -0,0 +1,431 @@
...
+             * There are no limitations on the size, as long as it fits a single
+             * layer of external local domains (near the global simulation area
+             * boundary) on each time step. In particular, PML is not required
```
[the remark this answers is not in the document]
Comment (sbastrakov) [reply include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp]: Agree this is confusing, will rephrase.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp [hunk 27]
```diff
@@ -0,0 +1,431 @@
...
+                 * at a border we set the corresponding thickness to 0.
+                 */
+                auto & movingWindow = MovingWindow::getInstance( );
+                auto const numSlides = movingWindow.getSlideCounter( currentStep );
+                auto const numExchanges = NumberOfExchanges< simDim >::value;
+                auto const communicationMask = Environment< simDim >::get( ).GridController( ).getCommunicationMask( );
+                Thickness localThickness = globalSize;
+                for( uint32_t exchange = 1u; exchange < numExchanges; ++exchange )
+                {
+                    // Skip directions except left, right, top, bottom, back, front
```
[the remark this answers is not in the document]
Comment (sbastrakov) [reply include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp]: Agree, will rephrase.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp [hunk 28]
```diff
@@ -0,0 +1,431 @@
...
+                    if( exchange >= BACK )
+                        axis = 2;
+                    uint32_t direction = exchange % 2;
+
+                    // No PML at the borders between two local domains
+                    bool hasNeighbour = communicationMask.isSet( exchange );
+                    if( hasNeighbour )
+                        localThickness( axis, direction ) = 0;
+
+                    // Disable PML during laser initialization
```
[the remark this answers is not in the document]
Comment (sbastrakov) [reply include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp]: Yes, in terms of interactions with the lasers' `initPlaneY` and the moving window PML mimics the other absorber.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp (hunk 10 above)
Comment (sbastrakov) [reply include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp]: Yes, these limitations will be lifted in following PRs. The plan for that is given in the first message in this discussion.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel (hunk 12 above)
Comment (sbastrakov) [reply include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel]: Will fix.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel (hunk 13 above)
Comment (sbastrakov) [reply include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel]: Sure.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel (hunk 14 above)
Comment (sbastrakov) [reply include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel]: Good point.

File: include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel (hunk 15 above)
Comment (sbastrakov) [reply include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel]: Will fix.

**Comment by sbastrakov** [comment]
@PrometheusPi thanks, I've pushed a new commit addressing all your comments. To repeat about the limitations (memory and checkpoint), they will be addressed in the following PRs.
